gateway/handler.go

```diff
@@ -623,6 +633,46 @@ func (h *handler) handleSubscribe(ctx context.Context, snSubscribe *snMsgs.Subsc
 	return h.mqttSend(mqSubscribe)
 }
 
+func (h *handler) handleUnsubscribe(ctx context.Context, snUnsubscribe *snMsgs.UnsubscribeMessage) error {
+	var topic string
+	var topicID uint16
+	switch snUnsubscribe.TopicIDType {
+	case snMsgs.TIT_STRING:
+		topic = string(snUnsubscribe.TopicName)
+
+		if !hasWildcard(topic) {
+			var found bool
+			topicID, found = h.findRegisteredTopicID(topic)
+			if !found {
+				snUnsuback := snMsgs.NewUnsubackMessage()
+				snUnsuback.CopyMessageID(snUnsubscribe)
+				return h.snSend(snUnsuback)
+			}
+		}
+
+		// topicID remains zero if client is unsubscribing from a wildcard topic.
+	case snMsgs.TIT_PREDEFINED:
+		var ok bool
+		topic, ok = h.predefinedTopics.GetTopicName(h.clientID, snUnsubscribe.TopicID)
+		if !ok {
+			return fmt.Errorf("Unknown topic id %d", snUnsubscribe.TopicID)
+		}
+		topicID = snUnsubscribe.TopicID
+	case snMsgs.TIT_SHORT:
+		topic = snMsgs.DecodeShortTopic(snUnsubscribe.TopicID)
+		// topicID remains zero.
+	}
```

Comment:
I think all this logic is relevant to `SUBSCRIBE` only, it's not applicable to `UNSUBSCRIBE`. Let's discuss it personally...